The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/10783034250/job/29904133557) [(plain)](https://github.com/rust-lang/rust/commit/21df264f31d575e9c420f27243abb254aab893b6/checks/29904133557/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
COPY host-x86_64/mingw-check/validate-error-codes.sh /scripts/

# NOTE: intentionally uses python2 for x.py so we can test it still works.
# validate-toolstate only runs in our CI, so it's ok for it to only support python3.
ENV SCRIPT TIDY_PRINT_DIFF=1 python2.7 ../x.py test \
           --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
# This file is autogenerated by pip-compile with Python 3.10
# by the following command:
#
#    pip-compile --allow-unsafe --generate-hashes reuse-requirements.in
---
#13 2.820 Building wheels for collected packages: reuse
#13 2.821   Building wheel for reuse (pyproject.toml): started
#13 3.063   Building wheel for reuse (pyproject.toml): finished with status 'done'
#13 3.064   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132715 sha256=dfa09868353292d98f811d3efdb0d54d07389e808efc71d68e3b93c514bf8bec
#13 3.064   Stored in directory: /tmp/pip-ephem-wheel-cache-4d56qm4k/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#13 3.067 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#13 3.452 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#13 3.453 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#13 3.979 Collecting virtualenv
#13 3.979 Collecting virtualenv
#13 4.032   Downloading virtualenv-20.26.4-py3-none-any.whl (6.0 MB)
#13 4.217      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.0/6.0 MB 33.0 MB/s eta 0:00:00
#13 4.260 Collecting distlib<1,>=0.3.7
#13 4.268   Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
#13 4.278      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 56.3 MB/s eta 0:00:00
#13 4.314 Collecting filelock<4,>=3.12.2
#13 4.322   Downloading filelock-3.16.0-py3-none-any.whl (16 kB)
#13 4.356 Collecting platformdirs<5,>=3.9.1
#13 4.364   Downloading platformdirs-4.3.2-py3-none-any.whl (18 kB)
#13 4.444 Installing collected packages: distlib, platformdirs, filelock, virtualenv
#13 4.632 Successfully installed distlib-0.3.8 filelock-3.16.0 platformdirs-4.3.2 virtualenv-20.26.4
#13 DONE 4.7s

#14 [7/8] COPY host-x86_64/mingw-check/validate-toolstate.sh /scripts/
#14 DONE 0.0s
---
DirectMap4k:      198592 kB
DirectMap2M:     9238528 kB
DirectMap1G:     9437184 kB
##[endgroup]
Executing TIDY_PRINT_DIFF=1 python2.7 ../x.py test            --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
+ TIDY_PRINT_DIFF=1 python2.7 ../x.py test --stage 0 src/tools/tidy tidyselftest --extra-checks=py:lint,cpp:fmt
    Finished `dev` profile [unoptimized] target(s) in 0.04s
##[endgroup]
downloading https://ci-artifacts.rust-lang.org/rustc-builds-alt/c2f74c3f928aeb503f15b4e9ef5778e77f3058b8/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz
extracting /checkout/obj/build/cache/llvm-c2f74c3f928aeb503f15b4e9ef5778e77f3058b8-true/rust-dev-nightly-x86_64-unknown-linux-gnu.tar.xz to /checkout/obj/build/x86_64-unknown-linux-gnu/ci-llvm
---
error: unexpected closing delimiter: `}`
##[error]  --> /checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_none_elf.rs:37:1
   |
33 |             ),
   |             - missing open `(` for this delimiter
37 | }
   | ^ unexpected closing delimiter

fmt: checked 5537 files
fmt: checked 5537 files
fmt error: Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/rustfmt/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/compiler/rustc_codegen_gcc/src/asm.rs" "/checkout/compiler/rustc_codegen_gcc/src/coverageinfo.rs" "/checkout/compiler/rustc_codegen_gcc/src/mono_item.rs" "/checkout/compiler/rustc_codegen_gcc/src/common.rs" "/checkout/compiler/rustc_codegen_gcc/src/int.rs" "/checkout/compiler/rustc_codegen_gcc/src/type_.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_be_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_uwp_windows_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64e_apple_ios.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparc64_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_musleabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32i_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_apple_ios_macabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_illumos.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7a_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_unknown_linux_gnu.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7a_none_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc64_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_freebsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/armeb_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imafc_unknown_nuttx_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_apple_watchos_sim.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7em_nuttx_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv5te_unknown_linux_gnueabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_wrs_vxworks_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv32imc_unknown_nuttx_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7m_nuttx_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm64ec_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7em_nuttx_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/wasm32_unknown_emscripten.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv7neon_linux_androideabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparc64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/powerpc_wrs_vxworks_spe.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv6k_nintendo_3ds.rs" "/checkout/compiler/rustc_target/src/spec/targets/xtensa_esp32_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7r_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_fuchsia.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_redox.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_uefi.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_rtems_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv6_unknown_netbsd_eabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/xtensa_esp32s2_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/i586_pc_nto_qnx700.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_none.rs" "/checkout/compiler/rustc_target/src/spec/targets/xtensa_esp32s3_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/sparcv9_sun_solaris.rs" "/checkout/compiler/rustc_target/src/spec/targets/armv7_unknown_linux_ohos.rs" "/checkout/compiler/rustc_target/src/spec/targets/i686_pc_windows_msvc.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_none_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/thumbv5te_none_eabi.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/mipsel_unknown_netbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/csky_unknown_linux_gnuabiv2hf.rs" "/checkout/compiler/rustc_target/src/spec/targets/aarch64_unknown_openbsd.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_linux_musl.rs" "/checkout/compiler/rustc_target/src/spec/targets/arm_unknown_linux_gnueabihf.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_nuttx_elf.rs" "/checkout/compiler/rustc_target/src/spec/targets/x86_64_fortanix_unknown_sgx.rs" "/checkout/compiler/rustc_target/src/spec/targets/riscv64gc_unknown_freebsd.rs" "/checkout/compiler/rustc_codegen_gcc/src/type_of.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
  local time: Mon Sep  9 23:48:03 UTC 2024
  network time: Mon, 09 Sep 2024 23:48:03 GMT
##[error]Process completed with exit code 1.
Post job cleanup.
```

</details>

--- a/compiler/rustc_codegen_llvm/src/context.rs
+++ b/compiler/rustc_codegen_llvm/src/context.rs
@@ -27,7 +27,7 @@ use rustc_span::source_map::Spanned;
 use rustc_span::{Span, DUMMY_SP};
 use rustc_target::abi::call::FnAbi;
 use rustc_target::abi::{HasDataLayout, TargetDataLayout, VariantIdx};
-use rustc_target::spec::{HasTargetSpec, RelocModel, Target, TlsModel};
+use rustc_target::spec::{HasTargetSpec, RelocModel, SmallDataThresholdSupport, Target, TlsModel};
 use smallvec::SmallVec;
 
 use crate::back::write::to_llvm_code_model;
@@ -386,6 +386,24 @@ pub(crate) unsafe fn create_module<'ll>(
             )
         }
     }
+
+    match (sess.opts.unstable_opts.small_data_threshold, sess.target.small_data_threshold_support())
+    {
+        // Set up the small-data optimization limit for architectures that use
+        // an LLVM module flag to control this.
+        (Some(threshold), SmallDataThresholdSupport::LlvmModuleFlag(flag)) => {
+            let flag = SmallCStr::new(flag.as_ref());
+            unsafe {
+                llvm::LLVMRustAddModuleFlagU32(
+                    llmod,
+                    llvm::LLVMModFlagBehavior::Error,
+                    flag.as_c_str().as_ptr(),
+                    threshold as u32,
+                )
+            }
+        }
+        _ => (),
+    };
 
     // Insert `llvm.ident` metadata.
     //

--- a/compiler/rustc_codegen_llvm/src/llvm_util.rs
+++ b/compiler/rustc_codegen_llvm/src/llvm_util.rs
@@ -14,7 +14,7 @@ use rustc_middle::bug;
 use rustc_session::config::{PrintKind, PrintRequest};
 use rustc_session::Session;
 use rustc_span::symbol::Symbol;
-use rustc_target::spec::{MergeFunctions, PanicStrategy};
+use rustc_target::spec::{MergeFunctions, PanicStrategy, SmallDataThresholdSupport};
 use rustc_target::target_features::{RUSTC_SPECIAL_FEATURES, RUSTC_SPECIFIC_FEATURES};
 
 use crate::back::write::create_informational_target_machine;
@@ -125,6 +125,18 @@ unsafe fn configure_llvm(sess: &Session) {
         for arg in sess_args {
             add(&(*arg), true);
         }
+
+        match (
+            sess.opts.unstable_opts.small_data_threshold,
+            sess.target.small_data_threshold_support(),
+        ) {
+            // Set up the small-data optimization limit for architectures that use
+            // an LLVM argument to control this.
+            (Some(threshold), SmallDataThresholdSupport::LlvmArg(arg)) => {
+                add(&format!("--{arg}={threshold}"), false)
+            }
+            _ => (),
+        };
     }
 
     if sess.opts.unstable_opts.llvm_time_trace {

--- a/compiler/rustc_interface/src/tests.rs
+++ b/compiler/rustc_interface/src/tests.rs
@@ -847,6 +847,7 @@ fn test_unstable_options_tracking_hash() {
     tracked!(share_generics, Some(true));
     tracked!(show_span, Some(String::from("abc")));
     tracked!(simulate_remapped_rust_src_base, Some(PathBuf::from("/rustc/abc")));
+    tracked!(small_data_threshold, Some(16));
     tracked!(split_lto_unit, Some(true));
     tracked!(src_hash_algorithm, Some(SourceFileHashAlgorithm::Sha1));
     tracked!(stack_protector, StackProtector::All);

--- a/compiler/rustc_session/messages.ftl
+++ b/compiler/rustc_session/messages.ftl
@@ -113,6 +113,8 @@ session_split_lto_unit_requires_lto = `-Zsplit-lto-unit` requires `-Clto`, `-Clt
 
 session_target_requires_unwind_tables = target requires unwind tables, they cannot be disabled with `-C force-unwind-tables=no`
 
+session_target_small_data_threshold_not_supported = `-Z small-data-threshold` is not supported for target {$target_triple} and will be ignored
+
 session_target_stack_protector_not_supported = `-Z stack-protector={$stack_protector}` is not supported for target {$target_triple} and will be ignored
 
 session_unleashed_feature_help_named = skipping check for `{$gate}` feature

--- a/compiler/rustc_session/src/errors.rs
+++ b/compiler/rustc_session/src/errors.rs
@@ -183,6 +183,12 @@ pub(crate) struct StackProtectorNotSupportedForTarget<'a> {
 }
 
 #[derive(Diagnostic)]
+#[diag(session_target_small_data_threshold_not_supported)]
+pub(crate) struct SmallDataThresholdNotSupportedForTarget<'a> {
+    pub(crate) target_triple: &'a TargetTriple,
+}
+
+#[derive(Diagnostic)]
 #[diag(session_branch_protection_requires_aarch64)]
 pub(crate) struct BranchProtectionRequiresAArch64;
 

--- a/compiler/rustc_session/src/options.rs
+++ b/compiler/rustc_session/src/options.rs
@@ -2014,6 +2014,8 @@ written to standard error output)"),
     simulate_remapped_rust_src_base: Option<PathBuf> = (None, parse_opt_pathbuf, [TRACKED],
         "simulate the effect of remap-debuginfo = true at bootstrapping by remapping path \
         to rust's source base directory. only meant for testing purposes"),
+    small_data_threshold: Option<usize> = (None, parse_opt_number, [TRACKED],
+        "Set the threshold for objects to be stored in a \"small data\" section"),
     span_debug: bool = (false, parse_bool, [UNTRACKED],
         "forward proc_macro::Span's `Debug` impl to `Span`"),
     /// o/w tests have closure@path

--- a/compiler/rustc_session/src/session.rs
+++ b/compiler/rustc_session/src/session.rs
@@ -30,8 +30,8 @@ use rustc_span::source_map::{FilePathMapping, SourceMap};
 use rustc_span::{FileNameDisplayPreference, RealFileName, Span, Symbol};
 use rustc_target::asm::InlineAsmArch;
 use rustc_target::spec::{
-    CodeModel, DebuginfoKind, PanicStrategy, RelocModel, RelroLevel, SanitizerSet, SplitDebuginfo,
-    StackProtector, Target, TargetTriple, TlsModel,
+    CodeModel, DebuginfoKind, PanicStrategy, RelocModel, RelroLevel, SanitizerSet,
+    SmallDataThresholdSupport, SplitDebuginfo, StackProtector, Target, TargetTriple, TlsModel,
 };
 
 use crate::code_stats::CodeStats;
@@ -1275,6 +1275,14 @@ fn validate_commandline_args_with_session_available(sess: &Session) {
                 stack_protector: sess.opts.unstable_opts.stack_protector,
                 target_triple: &sess.opts.target_triple,
             });
+        }
+    }
+
+    if sess.opts.unstable_opts.small_data_threshold.is_some() {
+        if sess.target.small_data_threshold_support() == SmallDataThresholdSupport::None {
+            sess.dcx().emit_warn(errors::SmallDataThresholdNotSupportedForTarget {
+                target_triple: &sess.opts.target_triple,
+            })
         }
     }
 

--- a/compiler/rustc_target/src/spec/mod.rs
+++ b/compiler/rustc_target/src/spec/mod.rs
@@ -855,6 +855,43 @@ impl ToJson for RelroLevel {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Hash)]
+pub enum SmallDataThresholdSupport {
+    None,
+    DefaultForArch,
+    LlvmModuleFlag(StaticCow<str>),
+    LlvmArg(StaticCow<str>),
+}
+
+impl FromStr for SmallDataThresholdSupport {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if s == "none" {
+            Ok(Self::None)
+        } else if s == "default-for-arch" {
+            Ok(Self::DefaultForArch)
+        } else if let Some(flag) = s.strip_prefix("llvm-module-flag=") {
+            Ok(Self::LlvmModuleFlag(flag.to_string().into()))
+        } else if let Some(arg) = s.strip_prefix("llvm-arg=") {
+            Ok(Self::LlvmArg(arg.to_string().into()))
+        } else {
+            Err(())
+        }
+    }
+}
+
+impl ToJson for SmallDataThresholdSupport {
+    fn to_json(&self) -> Value {
+        match self {
+            Self::None => "none".to_json(),
+            Self::DefaultForArch => "default-for-arch".to_json(),
+            Self::LlvmModuleFlag(flag) => format!("llvm-module-flag={flag}").to_json(),
+            Self::LlvmArg(arg) => format!("llvm-arg={arg}").to_json(),
+        }
+    }
+}
+
 #[derive(Clone, Copy, Debug, PartialEq, Hash)]
 pub enum MergeFunctions {
     Disabled,
@@ -2392,6 +2429,9 @@ pub struct TargetOptions {
 
     /// Whether the target supports XRay instrumentation.
     pub supports_xray: bool,
+
+    /// Whether the targets supports -Z small-data-threshold
+    small_data_threshold_support: SmallDataThresholdSupport,
 }
 
 /// Add arguments for the given flavor and also for its "twin" flavors
@@ -2609,6 +2649,7 @@ impl Default for TargetOptions {
             entry_name: "main".into(),
             entry_abi: Conv::C,
             supports_xray: false,
+            small_data_threshold_support: SmallDataThresholdSupport::DefaultForArch,
         }
     }
 }
@@ -2880,6 +2921,15 @@ impl Target {
                         _ => return Some(Err(format!("'{}' is not a valid TLS model. \
                                                       Run `rustc --print tls-models` to \
                                                       see the list of supported values.", s))),
+                    }
+                    Some(Ok(()))
+                })).unwrap_or(Ok(()))
+            } );
+            ($key_name:ident, SmallDataThresholdSupport) => ( {
+                obj.remove("small-data-threshold-support").and_then(|o| o.as_str().and_then(|s| {
+                    match s.parse::<SmallDataThresholdSupport>() {
+                        Ok(support) => base.small_data_threshold_support = support,
+                        _ => return Some(Err(format!("'{s}' is not a valid value for small-data-threshold-support."))),
                     }
                     Some(Ok(()))
                 })).unwrap_or(Ok(()))
@@ -3321,6 +3371,7 @@ impl Target {
         key!(supported_sanitizers, SanitizerSet)?;
         key!(generate_arange_section, bool);
         key!(supports_stack_protector, bool);
+        key!(small_data_threshold_support, SmallDataThresholdSupport)?;
         key!(entry_name);
         key!(entry_abi, Conv)?;
         key!(supports_xray, bool);
@@ -3413,6 +3464,30 @@ impl Target {
                 let obj = serde_json::from_str(contents).map_err(|e| e.to_string())?;
                 Target::from_json(obj)
             }
+        }
+    }
+
+    /// Return the target's small data threshold support, converting
+    /// `DefaultForArch` into a concrete value.
+    pub fn small_data_threshold_support(&self) -> SmallDataThresholdSupport {
+        match &self.options.small_data_threshold_support {
+            // Avoid having to duplicate the small data support in every
+            // target file by supporting a default value for each
+            // architecture.
+            SmallDataThresholdSupport::DefaultForArch => match self.arch.as_ref() {
+                "mips" | "mips64" | "mips32r6" => {
+                    SmallDataThresholdSupport::LlvmArg("mips-ssection-threshold".into())
+                }
+                "hexagon" => {
+                    SmallDataThresholdSupport::LlvmArg("hexagon-small-data-threshold".into())
+                }
+                "m68k" => SmallDataThresholdSupport::LlvmArg("m68k-ssection-threshold".into()),
+                "riscv32" | "riscv64" => {
+                    SmallDataThresholdSupport::LlvmModuleFlag("SmallDataLimit".into())
+                }
+                _ => SmallDataThresholdSupport::None,
+            },
+            s => s.clone(),
         }
     }
 }
@@ -3577,6 +3652,7 @@ impl ToJson for Target {
         target_option_val!(c_enum_min_bits);
         target_option_val!(generate_arange_section);
         target_option_val!(supports_stack_protector);
+        target_option_val!(small_data_threshold_support);
         target_option_val!(entry_name);
         target_option_val!(entry_abi);
         target_option_val!(supports_xray);

--- a/src/doc/unstable-book/src/compiler-flags/small-data-threshold.md
+++ b/src/doc/unstable-book/src/compiler-flags/small-data-threshold.md
@@ -1,0 +1,20 @@
+# `small-data-threshold`
+
+-----------------------
+
+This flag controls the maximum static variable size that may be included in the
+"small data sections" (.sdata, .sbss) supported by some architectures (RISCV,
+MIPS, M68K, Hexagon).  Can be set to `0` to disable the use of small data
+sections.
+
+Target support is indicated by the `small_data_threshold_support` target
+option which can be:
+
+- `none` (`SmallDataThresholdSupport::None`) for no support
+- `default-for-arch` (`SmallDataThresholdSupport::DefaultForArch`) which
+  is automatically translated into an appropriate value for the target.
+- `llvm-module-flag=<flag_name>`
+  (`SmallDataThresholdSupport::LlvmModuleFlag`) for specifying the
+  threshold via an LLVM module flag
+- `llvm-arg=<arg_name>` (`SmallDataThresholdSupport::LlvmArg`) for
+  specifying the threshold via an LLVM argument.

--- a/tests/assembly/small_data_threshold.rs
+++ b/tests/assembly/small_data_threshold.rs
@@ -1,0 +1,92 @@
+// Test for -Z small_data_threshold=...
+//@ revisions: RISCV MIPS HEXAGON M68K
+//@ assembly-output: emit-asm
+//@ compile-flags: -Z small_data_threshold=4
+//@ [RISCV] compile-flags: --target=riscv32im-unknown-none-elf
+//@ [RISCV] needs-llvm-components: riscv
+//@ [MIPS] compile-flags: --target=mips-unknown-linux-uclibc -C relocation-model=static
+//@ [MIPS] compile-flags: -C llvm-args=-mgpopt -C llvm-args=-mlocal-sdata
+//@ [MIPS] compile-flags: -C target-feature=+noabicalls
+//@ [MIPS] needs-llvm-components: mips
+//@ [HEXAGON] compile-flags: --target=hexagon-unknown-linux-musl -C target-feature=+small-data
+//@ [HEXAGON] compile-flags: -C llvm-args=--hexagon-statics-in-small-data
+//@ [HEXAGON] needs-llvm-components: hexagon
+//@ [M68K] compile-flags: --target=m68k-unknown-linux-gnu
+//@ [M68K] needs-llvm-components: m68k
+
+#![feature(no_core, lang_items)]
+#![no_std]
+#![no_core]
+#![crate_type = "lib"]
+
+#[lang = "sized"]
+trait Sized {}
+
+#[lang = "drop_in_place"]
+fn drop_in_place<T>(_: *mut T) {}
+
+#[used]
+#[no_mangle]
+// U is below the threshold, should be in sdata
+static mut U: u16 = 123;
+
+#[used]
+#[no_mangle]
+// V is below the threshold, should be in sbss
+static mut V: u16 = 0;
+
+#[used]
+#[no_mangle]
+// W is at the threshold, should be in sdata
+static mut W: u32 = 123;
+
+#[used]
+#[no_mangle]
+// X is at the threshold, should be in sbss
+static mut X: u32 = 0;
+
+#[used]
+#[no_mangle]
+// Y is over the threshold, should be in its own .data section
+static mut Y: u64 = 123;
+
+#[used]
+#[no_mangle]
+// Z is over the threshold, should be in its own .bss section
+static mut Z: u64 = 0;
+
+// Currently, only MIPS and RISCV successfully put any objects in the small data
+// sections so the U/V/W/X tests are skipped on Hexagon and M68K
+
+//@ RISCV: .section .sdata,
+//@ RISCV-NOT: .section
+//@ RISCV: U:
+//@ RISCV: .section .sbss
+//@ RISCV-NOT: .section
+//@ RISCV: V:
+//@ RISCV: .section .sdata
+//@ RISCV-NOT: .section
+//@ RISCV: W:
+//@ RISCV: .section .sbss
+//@ RISCV-NOT: .section
+//@ RISCV: X:
+
+//@ MIPS: .section .sdata,
+//@ MIPS-NOT: .section
+//@ MIPS: U:
+//@ MIPS: .section .sbss
+//@ MIPS-NOT: .section
+//@ MIPS: V:
+//@ MIPS: .section .sdata
+//@ MIPS-NOT: .section
+//@ MIPS: W:
+//@ MIPS: .section .sbss
+//@ MIPS-NOT: .section
+//@ MIPS: X:
+
+//@ CHECK: .section .data.Y,
+//@ CHECK-NOT: .section
+//@ CHECK: Y:
+//@ CHECK: .section .bss.Z,
+//@ CHECK-NOT: .section
+//@ CHECK: Z:

